stint/private/int_bitwise_ops.nim

```diff
@@ -24,3 +24,74 @@ func `and`*(x, y: IntImpl): IntImpl {.inline.}=
 func `xor`*(x, y: IntImpl): IntImpl {.inline.}=
   ## `Bitwise xor` of numbers x and y
   applyHiLo(x, y, `xor`)
+
+func `shr`*(x: IntImpl, y: SomeInteger): IntImpl {.inline.}
+  # Forward declaration
+
+func convertImpl[T: SomeInteger](x: SomeInteger): T {.compileTime.} =
+  cast[T](x)
+
+func convertImpl[T: IntImpl|UintImpl](x: IntImpl|UintImpl): T {.compileTime.} =
+  result.hi = convertImpl[type(result.hi)](x.hi)
+  result.lo = x.lo
+
+template convert[T](x: UintImpl|IntImpl|SomeInteger): T =
+  when nimvm:
+    # this is a workaround Nim VM inability to cast
+    # something non integer
+    convertImpl[T](x)
+  else:
+    cast[T](x)
+
+func `shl`*(x: IntImpl, y: SomeInteger): IntImpl {.inline.}=
+  ## Compute the `shift left` operation of x and y
+  # Note: inlining this poses codegen/aliasing issue when doing `x = x shl 1`
+
+  # TODO: would it be better to reimplement this with words iteration?
+  const halfSize: type(y) = bitsof(x) div 2
+  type HiType = type(result.hi)
+
+  if y == 0:
+    return x
+  elif y == halfSize:
+    result.hi = convert[HiType](x.lo)
+  elif y < halfSize:
+    result.hi = (x.hi shl y) or convert[HiType](x.lo shr (halfSize - y))
+    result.lo = x.lo shl y
+  else:
+    result.hi = convert[HiType](x.lo shl (y - halfSize))
+
+func `shr`*(x: IntImpl, y: SomeInteger): IntImpl {.inline.}=
+  ## Compute the `shift right` operation of x and y
+  ## Similar to C standard, result is undefined if y is bigger
+  ## than the number of bits in x.
+  const halfSize: type(y) = bitsof(x) div 2
+  type LoType = type(result.lo)
+
+  if y == 0:
+    return x
+  elif y == halfSize:
+    result.lo = convert[LoType](x.hi)
+  elif y < halfSize:
+    result.lo = (x.lo shr y) or convert[LoType](x.hi shl (halfSize - y))
+    result.hi = x.hi shr y
+  else:
+    result.lo = convert[LoType](x.hi shr (y - halfSize))
+
+func ashr*(x: IntImpl, y: SomeInteger): IntImpl {.inline.}=
+  ## Compute the `arithmetic shift right` operation of x and y
+  ## Similar to C standard, result is undefined if y is bigger
+  ## than the number of bits in x.
+  const halfSize: type(y) = bitsof(x) div 2
+  type LoType = type(result.lo)
+  if y == 0:
+    return x
+  elif y == halfSize:
+    result.lo = convert[LoType](x.hi)
+    result.hi = ashr(x.hi, halfSize-1)
+  elif y < halfSize:
+    result.lo = (x.lo shr y) or convert[LoType](x.hi shl (halfSize - y))
```

Comment:
are you sure we shouldn't `ashr` here?

Reply:
By definition, only `x.hi` part carries the `sign` bit. `x.lo` never have sign. And the `y` is not crossing the `hi-lo` boundary. It is safe to use only `shr` here.

for example, s = sign bit, z = hi bits, u = lo bits:

```text
      szzzzzzzuuuuuuuu

      sar 3                       # hi-lo boundary is 8, y is 3

      sssszzzzzzzuuuuu   # no need to use ashr for lo bits
```